File: Mac/JarvisMac/Logging/JarvisReplyLog.swift
import Foundation

// MARK: - Reply source

/// Where the spoken text originated from.
enum ReplySource: String, Codable {
    case command        // routed command response (renderer/ResponsePlaybook)
    case template       // explicit ResponseKey lookup
    case llm            // LLM-generated response
    case tool           // tool execution result
    case proactive      // unsolicited nudge or suggestion
    case error          // error/failure path
    case wake           // wake word acknowledgement
    case system         // internal system message (not editable)
}

// MARK: - Log entry

struct ReplyLogEntry: Codable, Identifiable {
    var id: UUID
    var timestamp: Date
    var source: ReplySource
    var intent: String?
    var responseKey: String?
    var spokenText: String
    var device: String
    var editableFlag: Bool          // false for system-only entries
    var correctedText: String?      // set when user edits via Reply Review
    var accepted: Bool?             // nil = not yet rated

    // Computed for export
    var isSystemOnly: Bool { !editableFlag }
}

// MARK: - Logger

/// Appends every spoken reply to a bounded JSONL file.
/// All reads and writes are on a dedicated serial queue — never call from main thread.
final class JarvisReplyLogger {

    static let shared = JarvisReplyLogger()

    private let fileURL: URL
    private let queue = DispatchQueue(label: "jarvis.reply-log", qos: .utility)
    private let maxBytes = 4 * 1024 * 1024  // 4 MB cap
    private let maxEntries = 2_000
    private var _cache: [ReplyLogEntry] = []
    private var _cacheLoaded = false

    private init() {
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = support.appendingPathComponent("JarvisMac", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        fileURL = dir.appendingPathComponent("reply_log.jsonl")
    }

    // MARK: Append

    func append(
        source: ReplySource,
        intent: String? = nil,
        responseKey: String? = nil,
        spokenText: String,
        editableFlag: Bool = true
    ) {
        guard !spokenText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        let entry = ReplyLogEntry(
            id: UUID(),
            timestamp: Date(),
            source: source,
            intent: intent,
            responseKey: responseKey,
            spokenText: spokenText,
            device: "mac",
            editableFlag: editableFlag,
            correctedText: nil,
            accepted: nil
        )
        queue.async { [weak self] in
            self?.write(entry: entry)
        }
    }

    private func write(entry: ReplyLogEntry) {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(entry),
              var line = String(data: data, encoding: .utf8) else { return }
        line += "\n"
        guard let lineData = line.data(using: .utf8) else { return }
        if !FileManager.default.fileExists(atPath: fileURL.path) {
            FileManager.default.createFile(atPath: fileURL.path, contents: nil)
        }
        if let handle = try? FileHandle(forWritingTo: fileURL) {
            handle.seekToEndOfFile()
            handle.write(lineData)
            try? handle.close()
        }
        // Append to in-memory cache if loaded
        if _cacheLoaded {
            _cache.append(entry)
            if _cache.count > maxEntries {
                _cache.removeFirst(_cache.count - maxEntries)
            }
        }
        rotateIfNeeded()
    }

    private func rotateIfNeeded() {
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: fileURL.path),
              let size = attrs[.size] as? Int,
              size > maxBytes else { return }
        // Keep second half of file
        guard let data = try? Data(contentsOf: fileURL),
              let content = String(data: data, encoding: .utf8) else { return }
        let lines = content.split(separator: "\n", omittingEmptySubsequences: true)
        let keep = Array(lines.dropFirst(lines.count / 2))
        let trimmed = keep.joined(separator: "\n") + "\n"
        try? trimmed.data(using: .utf8)?.write(to: fileURL, options: .atomic)
        _cacheLoaded = false  // invalidate cache
    }

    // MARK: Read

    func recentEntries(limit: Int = 100, completion: @escaping ([ReplyLogEntry]) -> Void) {
        queue.async { [weak self] in
            guard let self else { return }
            if !_cacheLoaded { _loadCache() }
            let result = Array(_cache.suffix(limit))
            DispatchQueue.main.async { completion(result) }
        }
    }

    func allEntries(completion: @escaping ([ReplyLogEntry]) -> Void) {
        queue.async { [weak self] in
            guard let self else { return }
            if !_cacheLoaded { _loadCache() }
            let result = _cache
            DispatchQueue.main.async { completion(result) }
        }
    }

    private func _loadCache() {
        guard let data = try? Data(contentsOf: fileURL),
              let content = String(data: data, encoding: .utf8) else {
            _cache = []; _cacheLoaded = true; return
        }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        _cache = content.split(separator: "\n", omittingEmptySubsequences: true).compactMap {
            try? decoder.decode(ReplyLogEntry.self, from: Data($0.utf8))
        }
        if _cache.count > maxEntries {
            _cache = Array(_cache.suffix(maxEntries))
        }
        _cacheLoaded = true
    }

    // MARK: Edit

    func updateCorrection(id: UUID, correctedText: String, accepted: Bool) {
        queue.async { [weak self] in
            guard let self else { return }
            if !_cacheLoaded { _loadCache() }
            if let idx = _cache.firstIndex(where: { $0.id == id }) {
                _cache[idx].correctedText = correctedText
                _cache[idx].accepted = accepted
                _rewriteFile()
            }
        }
    }

    private func _rewriteFile() {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        let lines = _cache.compactMap { entry -> String? in
            guard let data = try? encoder.encode(entry),
                  let s = String(data: data, encoding: .utf8) else { return nil }
            return s
        }
        let content = lines.joined(separator: "\n") + "\n"
        try? content.data(using: .utf8)?.write(to: fileURL, options: .atomic)
    }

    // MARK: Export

    /// Returns a JSONL string of all entries suitable for training export.
    func exportJSONL(completion: @escaping (String) -> Void) {
        allEntries { entries in
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            let lines = entries.compactMap { e -> String? in
                guard let data = try? encoder.encode(e),
                      let s = String(data: data, encoding: .utf8) else { return nil }
                return s
            }
            completion(lines.joined(separator: "\n"))
        }
    }

    /// Returns a CSV string of all entries.
    func exportCSV(completion: @escaping (String) -> Void) {
        allEntries { entries in
            var rows = ["timestamp,source,intent,responseKey,spokenText,correctedText,accepted,device,editableFlag"]
            for e in entries {
                let ts = ISO8601DateFormatter().string(from: e.timestamp)
                let row = [
                    ts,
                    e.source.rawValue,
                    e.intent ?? "",
                    e.responseKey ?? "",
                    Self.csvEscape(e.spokenText),
                    e.correctedText.map { Self.csvEscape($0) } ?? "",
                    e.accepted.map { $0 ? "true" : "false" } ?? "",
                    e.device,
                    e.editableFlag ? "true" : "false"
                ].joined(separator: ",")
                rows.append(row)
            }
            completion(rows.joined(separator: "\n"))
        }
    }

    private static func csvEscape(_ s: String) -> String {
        let escaped = s.replacingOccurrences(of: "\"", with: "\"\"")
        return "\"\(escaped)\""
    }
}

File: app/src/main/java/com/jarvis/assistant/logging/JarvisReplyLog.kt
package com.jarvis.assistant.logging

import android.content.Context
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.time.Instant
import java.util.UUID

// MARK: - Reply source

enum class ReplySource(val key: String) {
    COMMAND("command"),       // routed command response (PhraseRegistry)
    TEMPLATE("template"),     // explicit PhraseKey lookup
    LLM("llm"),               // LLM-generated response
    TOOL("tool"),             // tool execution result
    PROACTIVE("proactive"),   // unsolicited nudge or suggestion
    ERROR("error"),           // error / failure path
    WAKE("wake"),             // wake word acknowledgement
    SYSTEM("system");         // internal system message (not user-editable)

    companion object {
        fun fromKey(key: String): ReplySource = entries.firstOrNull { it.key == key } ?: SYSTEM
    }
}

// MARK: - Log entry

data class ReplyLogEntry(
    val id: String = UUID.randomUUID().toString(),
    val timestampMs: Long = Instant.now().toEpochMilli(),
    val source: String,
    val intent: String? = null,
    val phraseKey: String? = null,
    val spokenText: String,
    val device: String = "android",
    val editableFlag: Boolean = true,
    val correctedText: String? = null,
    val accepted: Boolean? = null,
)

// MARK: - Logger

/**
 * Appends every spoken reply to a bounded JSONL file at
 * [context.filesDir]/jarvis/reply_log.jsonl.
 *
 * Thread-safe: all I/O dispatched on [Dispatchers.IO].
 */
class JarvisReplyLogger private constructor(context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val logFile: File
    private val maxBytes = 4 * 1024 * 1024L   // 4 MB cap
    private val maxEntries = 2_000
    private val gson = Gson()

    private var cache: MutableList<ReplyLogEntry>? = null

    init {
        val dir = File(context.filesDir, "jarvis").also { it.mkdirs() }
        logFile = File(dir, "reply_log.jsonl")
    }

    companion object {
        @Volatile private var instance: JarvisReplyLogger? = null

        fun init(context: Context) {
            instance = JarvisReplyLogger(context.applicationContext)
        }

        fun get(): JarvisReplyLogger =
            instance ?: error("JarvisReplyLogger not initialized — call init() in Application.onCreate()")
    }

    // MARK: Append

    fun append(
        source: ReplySource,
        spokenText: String,
        intent: String? = null,
        phraseKey: String? = null,
        editableFlag: Boolean = true,
    ) {
        if (spokenText.isBlank()) return
        val entry = ReplyLogEntry(
            source      = source.key,
            intent      = intent,
            phraseKey   = phraseKey,
            spokenText  = spokenText,
            editableFlag = editableFlag,
        )
        scope.launch { writeEntry(entry) }
    }

    private fun writeEntry(entry: ReplyLogEntry) {
        val line = gson.toJson(entry) + "\n"
        logFile.appendText(line, Charsets.UTF_8)
        cache?.let {
            it.add(entry)
            if (it.size > maxEntries) repeat(it.size - maxEntries) { _ -> it.removeAt(0) }
        }
        rotateIfNeeded()
    }

    private fun rotateIfNeeded() {
        if (logFile.length() <= maxBytes) return
        val lines = logFile.readLines(Charsets.UTF_8).filter { it.isNotBlank() }
        val keep = lines.drop(lines.size / 2)
        logFile.writeText(keep.joinToString("\n") + "\n", Charsets.UTF_8)
        cache = null  // invalidate
    }

    // MARK: Read

    suspend fun recentEntries(limit: Int = 100): List<ReplyLogEntry> = withContext(Dispatchers.IO) {
        loadCache().takeLast(limit)
    }

    suspend fun allEntries(): List<ReplyLogEntry> = withContext(Dispatchers.IO) {
        loadCache().toList()
    }

    private fun loadCache(): MutableList<ReplyLogEntry> {
        cache?.let { return it }
        val loaded = if (!logFile.exists()) mutableListOf()
        else logFile.readLines(Charsets.UTF_8)
            .filter { it.isNotBlank() }
            .mapNotNull { line ->
                try { gson.fromJson(line, ReplyLogEntry::class.java) } catch (_: JsonSyntaxException) { null }
            }
            .toMutableList()
        if (loaded.size > maxEntries) loaded.subList(0, loaded.size - maxEntries).clear()
        cache = loaded
        return loaded
    }

    // MARK: Edit

    fun updateCorrection(id: String, correctedText: String, accepted: Boolean) {
        scope.launch {
            val entries = loadCache()
            val idx = entries.indexOfFirst { it.id == id }
            if (idx < 0) return@launch
            entries[idx] = entries[idx].copy(correctedText = correctedText, accepted = accepted)
            rewriteFile(entries)
        }
    }

    private fun rewriteFile(entries: List<ReplyLogEntry>) {
        val content = entries.joinToString("\n") { gson.toJson(it) } + "\n"
        logFile.writeText(content, Charsets.UTF_8)
    }

    // MARK: Export

    suspend fun exportJSONL(): String = withContext(Dispatchers.IO) {
        allEntries().joinToString("\n") { gson.toJson(it) }
    }

    suspend fun exportCSV(): String = withContext(Dispatchers.IO) {
        val header = "timestamp,source,intent,phraseKey,spokenText,correctedText,accepted,device,editableFlag"
        val rows = allEntries().map { e ->
            listOf(
                Instant.ofEpochMilli(e.timestampMs).toString(),
                e.source,
                e.intent.orEmpty(),
                e.phraseKey.orEmpty(),
                csvEscape(e.spokenText),
                e.correctedText?.let { csvEscape(it) }.orEmpty(),
                e.accepted?.toString().orEmpty(),
                e.device,
                e.editableFlag.toString(),
            ).joinToString(",")
        }
        (listOf(header) + rows).joinToString("\n")
    }

    private fun csvEscape(s: String) = "\"${s.replace("\"", "\"\"")}\""
}
